src/EditorFeatures/Core/Intents/RenameIntentProvider.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Features.Intents;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Rename;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EditorFeatures.Intents;
+
+[IntentProvider(WellKnownIntents.Rename, LanguageNames.CSharp), Shared]
+internal class RenameIntentProvider : IIntentProvider
+{
+    private record RenameIntentData(string NewName);
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public RenameIntentProvider()
+    {
+    }
+
+    public async Task<ImmutableArray<IntentProcessorResult>> ComputeIntentAsync(Document priorDocument, TextSpan priorSelection, Document currentDocument, IntentDataProvider intentDataProvider, CancellationToken cancellationToken)
+    {
+        var renameIntentData = intentDataProvider.GetIntentData<RenameIntentData>();
+        Contract.ThrowIfNull(renameIntentData);
+
+        var renameService = priorDocument.Project.LanguageServices.GetRequiredService<IEditorInlineRenameService>();
+        var renameInfo = await renameService.GetRenameInfoAsync(priorDocument, priorSelection.Start, cancellationToken).ConfigureAwait(false);
+        if (!renameInfo.CanRename)
+        {
+            return ImmutableArray<IntentProcessorResult>.Empty;
+        }
+
+        Contract.ThrowIfNull(renameIntentData);
+
+        var options = new SymbolRenameOptions(
+            RenameOverloads: false,
+            RenameInStrings: false,
+            RenameInComments: false,
+            RenameFile: false);
+
+        var renameLocationSet = await renameInfo.FindRenameLocationsAsync(options, cancellationToken).ConfigureAwait(false);
+        var renameReplacementInfo = await renameLocationSet.GetReplacementsAsync(renameIntentData.NewName, options, cancellationToken).ConfigureAwait(false);
```

Comment:
it's a pity this has to be computed up front.  It seems like we want a 'compute + resolve' approach where we determine if the rename is valid, then prose asks the user "do you want to rename?" then it calls back into us for the expensive bit.

renames (esp. on large solutions take a *lot* of time) so i'm worried this will basically make the feature not work unless you have small solutions.

Reply:
discussed offline - for now this is just for testing to make sure we can reliably detect this and give something valuable.  For actual release we will need to consider potentially different UX (for example one that allows us to lazily resolve the intent with something like a background work indicator).  This is applicable in general to all our code actions as they can be expensive to compute.  

There are some optimizations here we could consider - like for example providing the rename locations only in the current document upfront to intellicode (less expensive) and then only doing the full rename on commit of the action.  This is something we'll need to design (cc @pgroene)